### ISSUE TITLE
WithHigherPrecision does not work when used with MultiStatements

### ISSUE THREAD

**Issue by alexstoick** [snowflakedb/gosnowflake#816]
Snowflake Driver version: 1.6.20
Go Version + Arch: go version go1.20.4 darwin/arm64
Snowflake Server version: 7.19.2

---

Toggling `WithHigherPrecision` has no effect when using `MultiStatements`.

Code repro:

```golang
rowsWithPrecision, err := db.QueryContext(sf.WithHigherPrecision(context.Background()), "<statement>")

multiStatCtx, _ := sf.WithMultiStatement(context.Background(), 2)
multiRows, err := db.QueryContext(sf.WithHigherPrecision(multiStatCtx), "USE DATABASE x; <statement>");
```

If running the same statement in both cases, in the first one - numbers get returned formatted properly, in the 2nd one - numbers come back as strings.

I have done some investigation as to what might be causing this, and I think it's due to the way of fetching query results does **not** support the Arrow return type, and always returns JSON and as such no higher precision is available.

Running the query lands us here:
https://github.com/snowflakedb/gosnowflake/blob/master/connection.go#L356 - at this point `data.Data.QueryResultFormat` **is** `arrow`. Jumping through that function lands us, eventually here:
https://github.com/snowflakedb/gosnowflake/blob/master/monitoring.go#L232 - which calls: https://github.com/snowflakedb/gosnowflake/blob/master/monitoring.go#L226 - this calls `/queries/<qid>/result` endpoint which seems to be where the problem is - in the sense that this **always returns JSON**.

I'd appreciate any help/directions on how to unblock myself so I can _run multiple queries that return Arrow data-types_.

**Comment by sfc-gh-dszmolka** [comment]
hi and thank you for raising this issue with us. I'm trying to reproduce using your code snippet and the following runtime
`gosnowflake 1.6.20`
`go version go1.20.4 linux/amd64`

but I'm obviously doing something wrong because for both `WithHigherPrecision` and `WithMultiStatement + WithHigherPrecision` I'm getting the same result.

I did the following to try and set up the reproduction.

### 1. prepare Snowflake environment
```sql
insert into testnumber values (1.98904821670429939890482167042993);
select * from testnumber;
/*
C1
1.9890482167042993989048216704299300000
*/ --decimal is 32 long but precision in table is 37 long
```

### 2. full script i'm using
```go
package main

import (
  "context"
  "database/sql"
  "flag"
  "fmt"
  "log"
  "os"
  "strconv"

  sf "github.com/snowflakedb/gosnowflake"
)

func getDSN() (string, *sf.Config, error) {
  env := func(key string, failOnMissing bool) string {
    if value := os.Getenv(key); value != "" {
      return value
    }
    if failOnMissing {
      log.Fatalf("%v environment not set", key)
    }
    return ""
  }

  account := env("SNOWFLAKE_TEST_ACCOUNT", true)
  user := env("SNOWFLAKE_TEST_USER", true)
  password := env("SNOWFLAKE_TEST_PASSWORD", true)
  host := env("SNOWFLAKE_TEST_HOST", false)
  portStr := env("SNOWFLAKE_TEST_PORT", false)
  protocol := env("SNOWFLAKE_TEST_PROTOCOL", false)
  //database := env("SNOWFLAKE_TEST_DATABASE", true)
  //schema := env("SNOWFLAKE_TEST_SCHEMA", true)

  port := 443
  var err error
  if len(portStr) > 0 {
    port, err = strconv.Atoi(portStr)
    if err != nil {
      return "", nil, err
    }
  }

  cfg := &sf.Config{
    Account:  account,
    User:     user,
    Password: password,
    Host:     host,
    Port:     port,
    Protocol: protocol,
    //Database: database,
    //Schema:   schema,
  }

  dsn, err := sf.DSN(cfg)
  return dsn, cfg, err
}

func main() {
  if !flag.Parsed() {
    flag.Parse()
  }

  dsn, config, err := getDSN()
  if err != nil {
    log.Fatalf("Failed to create DSN from config: %v, error: %v", config, err)
  }

  db, err := sql.Open("snowflake", dsn)
  if err != nil {
    log.Fatalf("Failed to connect, dsn: %v, error: %v", dsn, err)
  }

  defer db.Close()

  query := "SELECT /* from-go */ c1 FROM TEST_DB.GO816.TESTNUMBER;"

  multiQuery := `
    USE SCHEMA TEST_DB.GO816;

    SELECT /* from-go */ c1 FROM TESTNUMBER;
  `
  // highprecision
  fmt.Println("Executing query WithHigherPrecision. \nResult on Snowflake GUI: 1.9890482167042993989048216704299300000")
  rowsWithPrecision, err := db.QueryContext(sf.WithHigherPrecision(context.Background()), query)
  if err != nil {
    log.Fatalf("Error while querying snowflake: %v", err)
  }

  defer rowsWithPrecision.Close()
  var v1 float64

  for rowsWithPrecision.Next() {
    err := rowsWithPrecision.Scan(&v1)
    if err != nil {
      log.Fatalf("Error while scanning row: %v", err)
    }
      fmt.Println("Result in gosnowflake:  ", v1)
  }
  if rowsWithPrecision.Err() != nil {
    log.Fatalf("Error while reading results: %v", err)
  }
  fmt.Println("=====")


  //multiquery + highprecision
  fmt.Println("Executing query WithHigherPrecision and WithMultiStatement. \nResult on Snowflake GUI: 1.9890482167042993989048216704299300000")
  multiStatCtx, _ := sf.WithMultiStatement(context.Background(), 2)
  multiRows, err := db.QueryContext(sf.WithHigherPrecision(multiStatCtx), multiQuery);

  if err != nil {
    log.Fatalf("Error while querying snowflake: %v", err)
  }

  defer multiRows.Close()
  var v2 float64

  for multiRows.NextResultSet() {
    for multiRows.Next() {
      err := multiRows.Scan(&v2)
      if err != nil {
        log.Fatalf("Error while scanning row: %v", err)
      }
      fmt.Println("Result in gosnowflake:  ", v2)
    }
  }
  if multiRows.Err() != nil {
    log.Fatalf("Error while reading results: %v", err)
  }

  fmt.Println("=====")
}
```

### 3. result:
```
# go run go816.go 
Executing query WithHigherPrecision. 
Result on Snowflake GUI: 1.9890482167042993989048216704299300000
Result in gosnowflake:   1.9890482167042993
=====
Executing query WithHigherPrecision and WithMultiStatement. 
Result on Snowflake GUI: 1.9890482167042993989048216704299300000
Result in gosnowflake:   1.9890482167042993
=====
```

so probably I'm missing something.

To investigate further, could you please provide the exact details such as
* the concrete number you're having problem with + Snowflake `NUMBER` column configuration
* the full runnable reproduction script, which when run, demonstrates the faulty behaviour

thank you in advance !

**Comment by sfc-gh-dszmolka** [comment]
got you - this detail helps, thanks! reproduced the issue successfully. we're taking a look.

as a workaround, if all you use the multistatement is to set the context (db / schema) for the table, you can use the fully qualified name and execute in a single statement. Except of course if the `"USE DATABASE x; <statement>"` was only for demonstration purposes; in this case I cannot provide any immediate workaround, except perhaps using some [casting like here](https://stackoverflow.com/questions/46783352/string-to-big-int-in-go) if that's possible.

**Comment by sfc-gh-dszmolka** [comment]
database is indeed an _optional_ configuration parameter for the connection and the session spawned from the connection inherits it, but this doesn't mean the session always has to work in the same context (db / schema context)

maybe a bit clumsy, but should perhaps be a viable workaround - to use the same connection but set the db/schema for the different queries, something like
```
setSchema, err := db.QueryContext(context, "USE SCHEMA DB1.SCHEMA1")
..do something with setSchema if necessary
result, err := db.QueryContext(context, "SELECT * FROM MY_ACTUAL_TABLE")
..handle result..
setAnotherSchema, err := db.QueryContext(context, "USE SCHEMA DB2.SCHEMA2")
..and so on
```
(please consider the above as a pseudo-code as it might contain syntax errors but hope I can get my point across)

As with the other drivers and even the GUI, these should execute in the same session, sequentially (and synchronously)
Checking the docs it suggests there's db.ExecContext() [to execute statements that don't return a result set](https://pkg.go.dev/github.com/snowflakedb/gosnowflake#section-readme) so perhaps could be used to execute the `USE DATABASE` / `USE SCHEMA` queries

For that matter, you can also use `db.Query()` instead to send in the `USE..` statements, since they are not expected to be cancellable or wanted to be cancelled. Actually, should be quite instantaneous. 

long story short: multi-statement is not the only option to execute multiple statements, they can be executed as single statements issued after each other in the same connection. Please test it and see if it works for you.

For the actual fix, I would like to hear back from the team first and see how they want to best address this, but of course the contributions are more than welcome, so if you already have some ideas , please go ahead and log a PR.

**Comment by sfc-gh-dszmolka** [comment]
The concern absolutely makes sense, yes, so recommend to test it to get proof if it's viable or not. I'll also share if there's other workarounds available.

**Comment by sfc-gh-dszmolka** [comment]
some update on this one. initial investigations show the driver works correctly, but the Snowflake backend itself returns data improperly. We're still looking if this is a bug or an undocumented feature. In the meantime a possible suggestion is to perhaps return the high precision values as strings, then cast them to `big.Float` on the application side.

anyhow, once there's more info we'll update the documentation accordingly and also update this Issue. thank you for bearing with us.

**Comment by sfc-gh-dszmolka** [comment]
on a totally unrelated thread i just learned that if one submits two goroutines from within one single Connection to do stuff, it actually creates two different Snowflake sessions. in their case, this is exactly what they wanted to avoid but since there was the concern from [the comment above](https://github.com/snowflakedb/gosnowflake/issues/816#issuecomment-1587699628) this might actually be an useful information for you or someone else stumbling into this one.

this is not related to the original issue, which is still under investigation why the backend returns data incorrectly.

**Comment by sfc-gh-dszmolka** [comment]
so after some further internal research, it turns out that it seems to work like this:

1. calling `Query`/`QueryContext` directly on `db` object - it creates a new Snowflake session every time
2. but if you create a connection directly, the session is reused, like this:

```go
db, err := sql.Open("snowflake", dsn)
	if err != nil {
		log.Fatalf("Failed to connect, dsn: %v, error: %v", dsn, err)
	}

	defer db.Close()

	ctx := context.Background()
	conn, err := db.Conn(ctx)
	defer conn.Close()

	conn.QueryContext(ctx, "SELECT 11")
	conn.QueryContext(ctx, "SELECT 12")
```

above 2 queries are executed as 2 separate (non multi-statement query) but in the same session

and now we're really far from the original issue, which is caused because of a limitation in the Snowflake backend. not being a `gosnowflake` driver problem, I'm now marking this issue as closed.
